Sage/carbon
# feat(scrollable-block): add parent prop and fix rendered HTML

ScrollableBlock currently renders as a pair of nested divs wrapping the list-item contents, which
produces invalid HTML. To fix this, ScrollableBlock will now render as a `<ul>` inside its own `<li>`.
Because this results in an unwanted additional nesting when the entire submenu is scrollable, the
outer `<ul>` is given role=presentation in this case so that screenreaders ignore the additional layer
of nesting. Also a parent prop has been added to ScrollableBlock to allow the `<li>` that wraps the
scrollable part to contain an item that acts as the parent of the whole scrollable sublist, which is
semantically more appropriate for some uses. A parentVariant prop has also been added to allow a
variant to be specified for this parent item.

### Proposed behaviour

ScrollableBlock should render valid HTML when used inside a submenu - in particular it should render as a `<li>` containing it's own `<ul>` wrapping the children (which are assumed to be MenuItem's which translate to HTML `<li>`'s).

ScrollableBlock should also accept a `parent` prop which accepts a React element which will be wrapped inside its own MenuItem and rendered inside the same `<li>` which holds the scrollable child `<ul>`, to allow a semantic relationship to be specified between that parent item and the scrollable list.



### Current behaviour

`ScrollableBlock` renders incorrect HTML with a pair of nested `<div>`s between the submenu's `<ul>` and the `<li>` children. This is invalid HTML, and even with a fix adding `role="presentation"`, the number of list items is announced incorrectly in some browser/screenreader combinations (notably Firefox).



### Checklist




#### QA


### Additional context



### Testing instructions



- ensure submenus, including scrollable submenus, behave as expected with a screenreader, across all browser/screenreader combinations
- test the following codesandbox, of a scrollable submenu with search, and ensure it is not broken after the PR: in particular that keyboard (arrow-key) navigation of the submenu still works: https://codesandbox.io/s/submenu-scrollable-search-working-example-i0tjcr
- test the new story "scrollable submenu with parent item", and ensure that works as the example above (including keyboard navigation), except for with a screenreader
- note the different in screenreader behaviour with the above two examples, and ensure that screenreaders make clear that the scrollable sublist is grouped with the search box in the same list item
